src/backend/distributed/utils/shardinterval_utils.c

```diff
@@ -297,7 +297,7 @@ FindShardIntervalIndex(Datum searchedValue, CitusTableCacheEntry *cacheEntry)
 	ShardInterval **shardIntervalCache = cacheEntry->sortedShardIntervalArray;
 	int shardCount = cacheEntry->shardIntervalArrayLength;
 	FmgrInfo *compareFunction = cacheEntry->shardIntervalCompareFunction;
-	bool useBinarySearch = (IsCitusTableTypeCacheEntry(cacheEntry, HASH_DISTRIBUTED) ||
+	bool useBinarySearch = (!IsCitusTableTypeCacheEntry(cacheEntry, HASH_DISTRIBUTED) ||
```

Comment:
minor: do you think would the following make it more explicit? Having multiple `!` along with `||` makes it a little hard for me to read the code :) 
 
bool findShardIndexForUnifiedRange = IsCitusTableTypeCacheEntry(cacheEntry, HASH_DISTRIBUTED) &&  cacheEntry->hasUniformHashDistribution;

```C
/* change the order of IF .. ELSE  as well*/
if (findShardIndexForUnifiedRange)
shardIndex = CalculateUniformHashRangeIndex(hashedValue, shardCount);
...
```

Feel free to ignore.

Reply:
I would then suggest to _not_ have this boolean, and instead use `cacheEntry->hasUniformHashDistribution` directly in the check where `useBinarySearch` currently is used.

This is in a branch that already verified the table is `HASH_DISTRIBUTED`, which AFAICT is the only use of this boolean.

The check for `cacheEntry->hasUniformHashDistribution` could have a explanation that explains we can use a fast path translation when the hash distribution has been uniform.